engine/runtime/src/main/java/org/enso/interpreter/runtime/ModuleSources.java

```diff
@@ -83,4 +87,51 @@ ModuleSources ensureSource(QualifiedName name) throws IOException {
   String getPath() {
     return file() == null ? null : file().getPath();
   }
+
+  /**
+   * Special support for reassigning locations of sources when running development version. When
+   * {@code -ea} is enabled, then we try to locate library files under the {@code
+   * distribution}/{@code lib} directories.
+   */
+  private static void alternativeFile(Source.SourceBuilder b, TruffleFile file) {
+    if (!BuildVersion.defaultDevEnsoVersion().equals(BuildVersion.ensoVersion())) {
+      // no changes when not running development version
+      return;
+    }
+    var root = file;
+    var stack = new LinkedList<String>();
+    while (root != null) {
+      if (BuildVersion.defaultDevEnsoVersion().equals(root.getName())) {
+        break;
+      }
+      stack.addFirst(root.getName());
+      root = root.getParent();
+    }
+    if (root == null) {
+      // give up when we cannot find a root
+      return;
+    }
```

Comment:
Perhaps we could do this root-location only once and then cache it?

Then we could log a single warning/info message about using the alternative locations. It is always useful to be able to tell which sources are used, so it would be great if this could tell us something in the logs.

Reply:
The code is also computing `stack` and that needs to run repeatably.

Reply:
I see. I was mostly suggesting 'caching' so that we can easily log just once.

I think it would be good to have a log message so that we know the location used is different from the typical one.